exec/internal/compile/amd64.go

```diff
@@ -0,0 +1,271 @@
+// Copyright 2017 The go-interpreter Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package compile
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+
+	ops "github.com/go-interpreter/wagon/wasm/operators"
+	asm "github.com/twitchyliquid64/golang-asm"
+	"github.com/twitchyliquid64/golang-asm/obj"
+	"github.com/twitchyliquid64/golang-asm/obj/x86"
+)
+
+// NativeCodeUnit represents compiled native code.
+type NativeCodeUnit interface {
+	Invoke(stack, locals *[]uint64)
+}
+
+// Details of the AMD64 backend:
+// Reserved registers (for now):
+//  - R10 - pointer to stack sliceHeader
+//  - R11 - pointer to locals sliceHeader
+//  - R12 - pointer for stack item
+//  - R13 - stack size
+// Scratch registers:
+//  - RAX, RBX, RCX, RDX, R8, R9, R14, R15
+// Most emission instructions make few attempts to optimize in order
+// to keep things simple, however a planned second pass peephole-optimizer
+//  should make a big difference.
+
+// AMD64Backend is the native compiler backend for x86-64 architectures.
+type AMD64Backend struct {
+	s *scanner
+}
+
+// Scanner returns a scanner that can be used for
+// emitting compilation candidates.
+func (b *AMD64Backend) Scanner() *scanner {
+	if b.s == nil {
+		b.s = &scanner{
+			supportedOpcodes: map[byte]bool{
+				ops.I64Const: true,
+				ops.I64Add:   true,
+				ops.I64Sub:   true,
+			},
+		}
+	}
+	return b.s
+}
+
+// Build implements exec.instructionBuilder.
+func (b *AMD64Backend) Build(candidate CompilationCandidate, code []byte, meta *BytecodeMetadata) ([]byte, error) {
+	builder, err := asm.NewBuilder("amd64", 64)
+	if err != nil {
+		return nil, err
+	}
+	b.emitPreamble(builder)
+
+	for i := candidate.StartInstruction; i <= candidate.EndInstruction; i++ {
+		//fmt.Printf("i=%d, meta=%+v, len=%d\n", i, meta.Instructions[i], len(code))
+		inst := meta.Instructions[i]
+		switch inst.Op {
+		case ops.I64Const:
+			immediate, err := b.readIntImmediate(code, inst)
+			if err != nil {
+				return nil, err
+			}
+			b.emitPushI64(builder, immediate)
+		case ops.I64Add, ops.I64Sub:
+			b.emitBinaryI64(builder, inst.Op)
+		default:
+			return nil, fmt.Errorf("cannot handle inst[%d].Op 0x%x", i, inst.Op)
+		}
+	}
+	ret := builder.NewProg()
+	ret.As = obj.ARET
+	builder.AddInstruction(ret)
+
+	out := builder.Assemble()
+	// cmd := exec.Command("ndisasm", "-b64", "-")
+	// cmd.Stdin = bytes.NewReader(out)
+	// cmd.Stdout = os.Stdout
+	// cmd.Run()
+	return out, nil
+}
+
+func (b *AMD64Backend) readIntImmediate(code []byte, meta InstructionMetadata) (uint64, error) {
+	if meta.Size == 5 {
+		var out uint32
+		err := binary.Read(bytes.NewReader(code[meta.Start+1:meta.Start+meta.Size]), binary.LittleEndian, &out)
```

Comment:
I'd use the faster `binary.LittleEndian.Uint32(code[...])` instead.
no need for an `error` in the signature, then.